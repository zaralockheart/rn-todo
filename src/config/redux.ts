// Imports: Dependencies
import {applyMiddleware, combineReducers, createStore} from 'redux'
import thunk from 'redux-thunk'
import AsyncStorage from '@react-native-community/async-storage'
import {createLogger} from 'redux-logger'
import {persistReducer, persistStore} from 'redux-persist'
import {reducer as formReducer} from 'redux-form'

// Imports: Reducers

// Redux: Root Reducer
const rootReducer = combineReducers({
	form: formReducer
});


// Middleware: Redux Thunk (Async/Await)
const middleware: any = [thunk];

// Middleware: Redux Logger (For Development)
if (process.env.NODE_ENV !== 'production') {
	middleware.push(createLogger());
}

// Middleware: Redux Persist Config
const persistConfig = {
	// Root?
	key: 'root',
	// Storage Method (React Native)
	storage: AsyncStorage,
	// Whitelist (Save Specific Reducers)
	whitelist: [
		'authReducer',
	],
	// Blacklist (Don't Save Specific Reducers)
	blacklist: [],
};

// Middleware: Redux Persist Persisted Reducer
const persistedReducer = persistReducer(persistConfig, rootReducer)

// Redux: Store
const store = createStore(
	persistedReducer,
	applyMiddleware(...middleware),
);

// Middleware: Redux Persist Persister
let persistor = persistStore(store);

// Exports
export {
	store,
	persistor,
};
