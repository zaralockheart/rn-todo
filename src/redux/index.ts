export * from './main'
export * from './todo'
