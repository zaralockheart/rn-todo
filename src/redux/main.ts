// Imports: Dependencies
import {applyMiddleware, combineReducers, createStore, Store} from 'redux'
import thunk from 'redux-thunk'
import AsyncStorage from '@react-native-community/async-storage'
import {createLogger} from 'redux-logger'
import {persistReducer, persistStore} from 'redux-persist'
import {FormStateMap, reducer as formReducer} from 'redux-form'
import {composeWithDevTools} from 'redux-devtools-extension'

import {todo, todoState} from './todo'

interface Actions<PROPS> {
    type: string
    payload: PROPS
}

interface ReducersAugment {
    form: FormStateMap
    todo: todoState
}

// Redux: Root Reducer
const rootReducer = combineReducers({
    form: formReducer,
    todo
})

// Middleware: Redux Thunk (Async/Await)
const middleware: [any] = [thunk]

// Middleware: Redux Logger (For Development)
if (process.env.NODE_ENV !== 'production') {
    middleware.push(createLogger())
}

// Middleware: Redux Persist Config
const persistConfig = {
    // Root?
    key: 'root',
    // Storage Method (React Native)
    storage: AsyncStorage,
    // Whitelist (Save Specific Reducers)
    whitelist: ['todo'],
    // Blacklist (Don't Save Specific Reducers)
    blacklist: []
}

// Middleware: Redux Persist Persisted Reducer
const persistedReducer = persistReducer(persistConfig, rootReducer)

// Redux: Store
export type State = ReducersAugment

const store: Store<State> = createStore(
    persistedReducer,
    composeWithDevTools(applyMiddleware(...middleware))
)

// Middleware: Redux Persist Persister
let persistor = persistStore(store)

// Exports
export {middleware, store, persistor, Actions, rootReducer}
