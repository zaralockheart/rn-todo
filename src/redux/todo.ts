import {Dispatch} from 'redux'
import uuidv5 from 'uuid/v5'

import {Actions} from './main'

// Action Names
const ADD_TODO = 'ADD_TODO'
const REMOVE_TODO = 'REMOVE_TODO'
const EDIT_TODO = 'EDIT_TODO'

// Actions
interface AddTodoProps {
  title: string
  subtitle?: string | undefined
  days: {[key: number]: string}
}

const addTodo = (data: AddTodoProps) => (dispatch: Dispatch) => {
  dispatch({
    type: ADD_TODO,
    payload: data
  })
}

interface EditTodoProps extends AddTodoProps {
  id: string
}

const editTodo = (data: EditTodoProps) => (dispatch: Dispatch) => {
  dispatch({
    type: EDIT_TODO,
    payload: data
  })
}

interface RemoveTodoProps {
  id: string
}

const removeTodo = (data: RemoveTodoProps) => (dispatch: Dispatch) => {
  dispatch({
    type: REMOVE_TODO,
    payload: data
  })
}

// All Props
type TodoProps = AddTodoProps & RemoveTodoProps

// Initial State
interface TodoObject {
  id: string
  title: string
  subtitle: string | undefined
  days: {[key: number]: string}
}

const initialState = {
  todo: {} as {[key: string]: TodoObject},
  ids: [] as string[]
}

type todoState = typeof initialState

const todo = (state = initialState, action: Actions<TodoProps>) => {
  switch (action.type) {
    case ADD_TODO: {
      const newId = !!state.ids ? state.ids.slice() : []
      let createdId = uuidv5(Date.now().toLocaleString(), uuidv5.DNS)
      newId.push(createdId)
      const newTodo: {[key: string]: TodoObject} = {
        ...state.todo
      }

      newTodo[createdId] = {
        id: createdId,
        subtitle: action.payload.subtitle,
        title: action.payload.title,
        days: action.payload.days
      }

      const newState: todoState = {
        ids: newId,
        todo: newTodo
      }
      return newState
    }
    case REMOVE_TODO: {
      const newIds = state.ids.slice().filter(i => i !== action.payload.id)

      const newTodo = {
        ...state.todo
      }

      delete newTodo[action.payload.id]
      const newState: todoState = {
        ids: newIds,
        todo: newTodo
      }
      return newState
    }
    case EDIT_TODO: {
      const newTodo = {
        ...state.todo
      }

      newTodo[action.payload.id] = {
        id: action.payload.id,
        title: action.payload.title,
        subtitle: action.payload.subtitle,
        days: action.payload.days
      }

      const newState: todoState = {
        ids: state.ids,
        todo: newTodo
      }

      return newState
    }
    default:
      return state
  }
}

export {TodoProps, TodoObject, todoState, todo, addTodo, removeTodo, editTodo}
