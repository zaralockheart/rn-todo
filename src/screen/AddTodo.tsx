import React from 'react'
import {SafeAreaView, StyleSheet, Text, TouchableOpacity, View} from 'react-native'
import {ErrorOther, Field, InjectedFormProps, reduxForm} from 'redux-form'
import _ from 'lodash'
import {useNavigation, useNavigationParam} from 'react-navigation-hooks'

import {Input, ListSelector} from '../components/form'
import {useDispatch} from 'react-redux'
import {addTodo, editTodo} from '../redux'
import {isNotEmpty} from '../util/main'

export const days = ['MON', 'TUE', 'WED', 'THU', 'FRI', 'SAT', 'SUN']

interface TodoFormProps {
  title: string
  subtitle: string
  days: { [key: number]: string }
}

const AddTodo = (props: InjectedFormProps<TodoFormProps>) => {
  const {goBack} = useNavigation()

  const dispatch = useDispatch()

  const id = useNavigationParam('id')

  const save = (obj: TodoFormProps) => {
    if (id) {
      dispatch(
        editTodo({
          id,
          days: obj.days,
          title: obj.title,
          subtitle: obj.subtitle
        })
      )
    } else {
      dispatch(
        addTodo({
          days: obj.days,
          title: obj.title,
          subtitle: obj.subtitle
        })
      )
    }

    goBack()
  }

  return (
    <SafeAreaView style={{flex: 1}}>
      <View style={styles.parent}>
        <Text>Title</Text>
        <Field name="title" component={Input} placeholder={'What is is about?'}/>
        <Text>Subtitle</Text>
        <Field
          name="subtitle"
          component={Input}
          placeholder={'Any more details?'}
        />
        <Field
          name="days"
          component={ListSelector}
          items={days.map((day, index) => ({
            id: index,
            text: day
          }))}
        />
        <View style={{flex: 1}}/>
        <TouchableOpacity
          style={styles.save}
          disabled={props.invalid}
          onPress={props.handleSubmit(save)}>
          <Text>Save</Text>
        </TouchableOpacity>
      </View>
    </SafeAreaView>
  )
}

const AddTodoForm = reduxForm<TodoFormProps>({
  form: 'TODO_FORM',
  enableReinitialize: true,
  validate(
    values: TodoFormProps
  ): { [P in keyof TodoFormProps]?: React.ReactElement | string } &
    ErrorOther<string> {
    const errors: Partial<{ [x in keyof TodoFormProps]: string }> = {
      title: undefined,
      days: undefined
    }

    errors.title = isNotEmpty(values.title, 'Title is required')
    errors.days = _.isEmpty(values.days)
      ? 'Required at least on day'
      : undefined

    return errors
  }
})(AddTodo)

export default () => {
  const subtitle = useNavigationParam('subtitle')
  const title = useNavigationParam('title')
  const days = useNavigationParam('days')

  return (
    <AddTodoForm
      initialValues={{
        title,
        subtitle,
        days
      }}
    />
  )
}

const styles = StyleSheet.create({
  parent: {
    flex: 1,
    padding: 10,
  },
  save: {
    backgroundColor: 'brown',
    padding: 20,
    alignItems: 'center',
    borderRadius: 10,
  }
})
