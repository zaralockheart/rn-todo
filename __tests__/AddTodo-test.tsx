/**
 * @format
 * @jest-environment jsdom
 */

import React from 'react'
import 'react-native'
import {mount} from 'enzyme'
import {combineReducers, createStore} from 'redux'
import {Provider} from 'react-redux'
import {Field, reducer as formReducer} from 'redux-form'

import AddTodo from '../src/screen/AddTodo'
import renderer from "react-test-renderer";

const rootReducer = combineReducers({
  form: formReducer
})

const store = createStore(rootReducer)

describe('Add Todo Form', () => {
  it('renders correctly', () => {
    const component = renderer.create(
      <Provider store={store}>
        <AddTodo />
      </Provider>
    )

    expect(component.toJSON()).toMatchSnapshot()
  })
})
