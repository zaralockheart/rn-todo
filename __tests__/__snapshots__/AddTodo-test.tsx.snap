// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Todo Form renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <View>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="What is is about?"
      rejectResponderTermination={true}
      style={
        Object {
          "borderColor": "black",
          "borderWidth": 1,
          "height": 37,
          "width": 250,
        }
      }
      testID="title"
      underlineColorAndroid="transparent"
      value=""
    />
    <Text>
      Title is required
    </Text>
  </View>
  <View>
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="Any more details?"
      rejectResponderTermination={true}
      style={
        Object {
          "borderColor": "black",
          "borderWidth": 1,
          "height": 37,
          "width": 250,
        }
      }
      testID="subtitle"
      underlineColorAndroid="transparent"
      value=""
    />
    <Text />
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 100,
          "height": 50,
          "justifyContent": "center",
          "opacity": 1,
          "padding": 5,
          "width": 50,
        }
      }
    >
      <Text>
        MON
      </Text>
    </View>
    <View
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 100,
          "height": 50,
          "justifyContent": "center",
          "opacity": 1,
          "padding": 5,
          "width": 50,
        }
      }
    >
      <Text>
        TUE
      </Text>
    </View>
    <View
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 100,
          "height": 50,
          "justifyContent": "center",
          "opacity": 1,
          "padding": 5,
          "width": 50,
        }
      }
    >
      <Text>
        WED
      </Text>
    </View>
    <View
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 100,
          "height": 50,
          "justifyContent": "center",
          "opacity": 1,
          "padding": 5,
          "width": 50,
        }
      }
    >
      <Text>
        THU
      </Text>
    </View>
    <View
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 100,
          "height": 50,
          "justifyContent": "center",
          "opacity": 1,
          "padding": 5,
          "width": 50,
        }
      }
    >
      <Text>
        FRI
      </Text>
    </View>
    <View
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 100,
          "height": 50,
          "justifyContent": "center",
          "opacity": 1,
          "padding": 5,
          "width": 50,
        }
      }
    >
      <Text>
        SAT
      </Text>
    </View>
    <View
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 100,
          "height": 50,
          "justifyContent": "center",
          "opacity": 1,
          "padding": 5,
          "width": 50,
        }
      }
    >
      <Text>
        SUN
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    clickable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <Text>
      Save
    </Text>
  </View>
</RCTSafeAreaView>
`;
